ERCS/erc-7674.md

```diff
@@ -0,0 +1,58 @@
+---
+eip: 7674
+title: Temporary Approval Extension for ERC-20
+description: An interface for ephemeral ERC-20 approvals
+author: Xenia Shape (@byshape), Mikhail Melnik (@ZumZoom), Hadrien Croubois (@Amxx)
+discussions-to: https://ethereum-magicians.org/t/add-erc-7674-transient-approval-extension-for-erc-20/19521
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-04-02
+requires: 20, 1153
+---
+
+## Abstract
+
+This specification defines the minimum interface required to temporarily approve [ERC-20](./eip-20.md) tokens for spending within the same transaction.
+
+## Motivation
+
+User are often require to set token approval that will only be used once. It is common to leave unexpected approvals after these interactions. [EIP-1153](./eip-1153.md) introduces a cheap way to handle temporarily allowances.
+
+## Specification
+
+The key words "MUST", "SHOULD", "MAY" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+Compliant contracts MUST implement 1 new function in addition to [ERC-20](./eip-20.md):
+```solidity
+function temporaryApprove(address spender, uint256 value) public returns (bool success)
+```
+Call to `temporaryApprove(spender, value)` allows `spender` to withdraw within the same transaction from `msg.sender` multiple times, up to the `value` amount.  This temporary allowance is to be considered in addition to the normal (persistent) [ERC-20](./eip-20.md) allowance, the total value that spender is able to spend during the transaction is thus capped by the sum of the temporary and the normal (persistent) allowances.
```

Comment:
>This temporary allowance is to be considered in addition to the normal (persistent) ERC-20 allowance

This forces an SLOAD of the current allowance, resulting in higher costs than one would want from this feature. I think it should be a temporary allowance override instead.

Reply:
I disagree. 

If the temporary allowance takes priority (this should be mentionned somewhere if it isn't already), and if the temporary allowance is sufficient to cover the need of some function that consumes allowance (`transferFrom`), then the allowance can be consumed from the temporary part without having to go consume anything in the persistent part.

You only need to lean (and update) the persistent part if the temporary part is not sufficient.

An analogy would be:
- you want to cook rice,
- you have 2kg of rice in your kitchen (fast access - eq to temporary)
- you have 50kg of rice stored in you basement (slow access - eq to persistent)

it is true that:
- you have 52kg available
- to measure how much you have available, you have to check both the kitchen and the basement
- if you need to cook diner, and that diner requires 500g of rice, the kitchen is sufficient, and you don't need to go to the basement. Rice in the basement is available if you need, but you don't need to access it unless there isn't enough in the kitchen for what you are planning to cook.

Reply:
What do you think of

```suggestion
Call to `temporaryApprove(spender, value)` allows `spender` to withdraw within the same transaction from `msg.sender` multiple times, up to the `value` amount.  This temporary allowance is to be considered in addition to the normal (persistent) [ERC-20](./eip-20.md) allowance, the total value that spender is able to spend during the transaction is thus capped by the sum of the temporary and the normal (persistent) allowances. 

While it SHOULD be possible for a `transferFrom` operation to consume both time of allowance, the consumption of the temporary allowance SHOULD take priority over the consumption of the persistent allowance. Therefore, if the temporary allowance is sufficient for executing a `transferFrom` operation, the persistent allowance SHOULD not be loaded/updated from the storage. Consumption of persistent allowance, which implies storage accesses, SHOULD be performed only if the temporary allowance is not sufficient for the operation being executed.
```

Reply:
Yes, my confusion was because of the effect on the `allowance()` function. If the allowances have to be added, this forces `allowance` to read from both. However, in the context of `transferFrom` it should not be necessary to do this, as you've described.

I am not sure if the effect on `allowance` is something that we should care about.